modules/wallet/files.go

```diff
@@ -1,112 +1,80 @@
 package wallet
 
 import (
-	// "io/ioutil"
-	// "os"
+	"errors"
+	"io/ioutil"
+	"os"
 
 	"github.com/NebulousLabs/Sia/consensus"
 	"github.com/NebulousLabs/Sia/crypto"
-	// "github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/encoding"
 )
 
-// Key is how we serialize and store spendable addresses on disk.
-//
-// TODO: When recreating this file, perhaps there are more intelligent things
-// that can be done than just saving the keys. Otherwise at start the wallet
-// needs to do a full catchup on the state, which involves requesting the diffs
-// of every single block.
-type Key struct {
-	UnlockConditions consensus.UnlockConditions
+func fileExists(filename string) bool {
+	_, err := os.Stat(filename)
+	return !os.IsNotExist(err)
+}
+
+type savedKey struct {
 	SecretKey        crypto.SecretKey
+	UnlockConditions consensus.UnlockConditions
 }
 
+// save writes the contents of a wallet to a file.
 func (w *Wallet) save() (err error) {
-	/*
-		// Add every known spendable address + secret key.
-		var i int
-		keys := make([]Key, len(w.addresses))
-		for _, address := range w.addresses {
-			key := Key{
-				SpendConditions: address.spendConditions,
-				SecretKey:       address.secretKey,
-			}
-			keys[i] = key
-			i++
-		}
-
-		timelockedKeys := make([]AddressKey, len(w.timelockedAddresses))
-		for _, address := range w.timelockedAddresses {
-			key := Key{
-				SpendConditions: address.spendConditions,
-				SecretKey:       address.secretKey,
-			}
-			timelockedKeys[i] = key
-			i++
-		}
+	// convert key map to a slice
+	keySlice := make([]savedKey, len(w.keys))
+	for _, key := range w.keys {
+		keySlice = append(keySlice, savedKey{key.secretKey, key.unlockConditions})
+	}
 
-		// Write the file.
-		fileData := encoding.MarshalAll(keys, timelockedKeys)
-		if err != nil {
-			return
-		}
-		// TODO: Instead of using WriteFile, write to a temp file and then do a move,
-		// this is an action that should probably appear in a library somewhere.
-		err = ioutil.WriteFile(w.saveFilename, fileData, 0666)
-		if err != nil {
-			return
-		}
-	*/
+	// Write the data to file.
+	// TODO: Instead of using WriteFile, write to a temp file and then do a move.
+	// TODO: append instead of overwriting everything.
```

Comment:
Append might be tricky because you have to have some sort of awareness of which keys are already in the file. What we'd probably do instead is call a saveSingleKey() function to write to the temp file, and then a mergeTemp() function that takes the temp file and the full file and puts them together.

Or if you get appending working, saveSingleKey() can just do the append.

I don't consider it necessary to make these a TODO, since there's already an issue for it.